### PR TITLE
Use ARC over MRC

### PR DESCRIPTION
**Pull request by Jamminroot**
This PR aims to replace manual reference counting with automatic reference counting

AB#79694

### CODE REVIEW

File: Other Sources/Crash Demo/main.m [hunk 1]
```diff
@@ -27,7 +27,7 @@
  */
 
 #import <Foundation/Foundation.h>
-#import "CrashReporter.h"
+#import "../../Source/CrashReporter.h"
```
Comment (MatkovIvan) [inline Other Sources/Crash Demo/main.m:30]: Should be reverter

File: Other Sources/Crash Demo/main.m [hunk 2]
```diff
@@ -136,7 +136,6 @@ static bool debugger_should_exit (void) {
 }
 
 int main (int argc, char *argv[]) {
-    NSAutoreleasePool *pool = [[NSAutoreleasePool alloc] init];
```
Comment (MatkovIvan) [inline Other Sources/Crash Demo/main.m:139]: `@autoreleasepool`

File: Other Sources/plcrashutil/main.m [hunk 3]
```diff
@@ -117,7 +117,6 @@ static int convert_command (int argc, char *argv[]) {
 }
 
 int main (int argc, char *argv[]) {
-    NSAutoreleasePool *pool = [[NSAutoreleasePool alloc] init];
```
Comment (MatkovIvan) [inline Other Sources/plcrashutil/main.m:120]: `@autoreleasepool`

File: Source/PLCrashReport.m [hunk 4]
```diff
@@ -114,75 +114,74 @@ - (id) initWithData: (NSData *) encodedData error: (NSError **) outError {
...
-    [_exceptionInfo release];
+    _systemInfo = NULL;
+    _machineInfo = NULL;
+    _applicationInfo = NULL;
+    _processInfo = NULL;
+    _signalInfo = NULL;
+    _machExceptionInfo = NULL;
+    _threads = NULL;
+    _images = NULL;
+    _exceptionInfo = NULL;
```
Comment (MatkovIvan) [inline Source/PLCrashReport.m:184]: I guess no need for it. Also, please use `nil` in ObjC code

File: Source/PLCrashReportExceptionInfo.m [hunk 5]
```diff
@@ -58,19 +58,11 @@ - (id) initWithExceptionName: (NSString *) name reason: (NSString *) reason stac
     if ((self = [super init]) == nil)
         return nil;
     
-    _name = [name retain];
-    _reason = [reason retain];
-    _stackFrames = [stackFrames retain];
+    _name = name;
+    _reason = reason;
+    _stackFrames = stackFrames;
```
Comment (MatkovIvan) [inline Source/PLCrashReportExceptionInfo.m:63]: Add explicit modifiers in header

File: Source/PLCrashReportTextFormatter.m [hunk 6]
```diff
@@ -182,8 +182,7 @@ + (NSString *) stringValueForCrashReport: (PLCrashReport *) report withTextForma
 
         NSString *incidentIdentifier = @"???";
         if (report.uuidRef != NULL) {
-            incidentIdentifier = (NSString *) CFUUIDCreateString(NULL, report.uuidRef);
-            [incidentIdentifier autorelease];
+            incidentIdentifier = (NSString *) CFBridgingRelease(CFUUIDCreateString(NULL, report.uuidRef));
```
Comment (MatkovIvan) [inline Source/PLCrashReportTextFormatter.m:185]: Use `__bridge_transfer`

File: Source/PLCrashReporter.m [hunk 7]
```diff
@@ -54,6 +54,12 @@
     NSLog(@"[PLCrashReporter] " msg, ## args); \
 }
 
+/**
+ * Static fields to prevent ARC from freeing resuources. */
+static PLCrashMachExceptionServer *mach_server;
+static PLCrashMachExceptionPortSet *port_set;
```
Comment (MatkovIvan) [inline Source/PLCrashReporter.m:60]: Remove it

File: Source/PLCrashReporter.m [hunk 8]
```diff
@@ -54,6 +54,12 @@
     NSLog(@"[PLCrashReporter] " msg, ## args); \
 }
 
+/**
+ * Static fields to prevent ARC from freeing resuources. */
```
Comment (MatkovIvan) [inline Source/PLCrashReporter.m:58]: Format

File: Source/PLCrashReporter.m [hunk 9]
```diff
@@ -608,24 +614,27 @@ - (BOOL) enableCrashReporterAndReturnError: (NSError **) outError {
                 return NO;
             
             /* Enable the server. */
-            _machServer = [self enableMachExceptionServerWithPreviousPortSet: &_previousMachPorts
+            /* TODO review prevMacPorts usage */
```
Comment (MatkovIvan) [inline Source/PLCrashReporter.m:617]: TODO

File: Source/PLCrashReporter.m [hunk 10]
```diff
@@ -608,24 +614,27 @@ - (BOOL) enableCrashReporterAndReturnError: (NSError **) outError {
...
+            port_set = _previousMachPorts;
             
             /*
              * MEMORY WARNING: To ensure that our instance survives for the lifetime of the callback registration,
              * we retain it here. This is necessary to ensure that the Mach exception server instance and previous port set
              * survive for the lifetime of the callback. Since there's currently no support for *deregistering* a crash reporter,
              * this simply results in the reporter living forever.
              */
-            [self retain];
+            pl_crash_reporter = self;
```
Comment (MatkovIvan) [inline Source/PLCrashReporter.m:637]: Update comment and take care about overwriting

File: Tests/PLCrashAsyncMachOImageTests.m [hunk 11]
```diff
@@ -48,7 +48,7 @@ @implementation PLCrashAsyncMachOImageTests
 - (void) setUp {
     /* Fetch our containing image's dyld info */
     Dl_info info;
-    STAssertTrue(dladdr([self class], &info) > 0, @"Could not fetch dyld info for %p", [self class]);
+    STAssertTrue(dladdr((__bridge_retained void *)([self class]), &info) > 0, @"Could not fetch dyld info for %p", [self class]);
```
Comment (MatkovIvan) [inline Tests/PLCrashAsyncMachOImageTests.m:51]: Not sure that `__bridge_retained` usage is correct here

File: Tests/PLCrashAsyncTests.m [hunk 12]
```diff
@@ -62,7 +62,7 @@ - (void) tearDown {
 
     /* Delete the file */
     STAssertTrue([[NSFileManager defaultManager] removeItemAtPath: _outputFile error: &error], @"Could not remove log file");
-    [_outputFile release];
+    _outputFile = NULL;
```
Comment (MatkovIvan) [inline Tests/PLCrashAsyncTests.m:65]:
```suggestion
    _outputFile = nil;
```

File: Tests/PLCrashHostInfoTests.m [hunk 13]
```diff
@@ -43,11 +43,11 @@ @interface PLCrashHostInfoTests : SenTestCase {
 @implementation PLCrashHostInfoTests
 
 - (void) setUp {
-    _hostInfo = [[PLCrashHostInfo currentHostInfo] retain];
+    _hostInfo = [PLCrashHostInfo currentHostInfo];
 }
 
 - (void) tearDown {
-    [_hostInfo release];
+    _hostInfo = NULL;
```
Comment (MatkovIvan) [inline Tests/PLCrashHostInfoTests.m:50]:
```suggestion
    _hostInfo = nil;
```

File: Source/PLCrashMachExceptionPortSet.m [hunk 14]
```diff
@@ -106,20 +106,14 @@ - (id) initWithAsyncSafeRepresentation: (plcrash_mach_exception_port_set_t) asyn
...
 
-- (void) dealloc {
-    [_state_set release];
-    [super dealloc];
-}
-
-
 // from NSFastEnumeration protocol
-- (NSUInteger) countByEnumeratingWithState: (NSFastEnumerationState *) state objects: (id *) stackbuf count: (NSUInteger) len {
+- (NSUInteger) countByEnumeratingWithState: (NSFastEnumerationState *) state objects: (id __unsafe_unretained _Nullable [_Nonnull]) stackbuf count: (NSUInteger) len {
```
Comment (MatkovIvan) [inline Source/PLCrashMachExceptionPortSet.m:116]: What does `_Nullable [_Nonnull]` mean?

File: Source/PLCrashReportApplicationInfo.m [hunk 15]
```diff
@@ -49,20 +49,13 @@ - (id) initWithApplicationIdentifier: (NSString *) applicationIdentifier
     if ((self = [super init]) == nil)
         return nil;
 
-    _applicationIdentifier = [applicationIdentifier retain];
-    _applicationVersion = [applicationVersion retain];
-    _applicationMarketingVersion = [applicationMarketingVersion retain];
+    _applicationIdentifier = applicationIdentifier;
+    _applicationVersion = applicationVersion;
+    _applicationMarketingVersion = applicationMarketingVersion;
```
Comment (MatkovIvan) [inline Source/PLCrashReportApplicationInfo.m:54]: Add explicit modifiers

File: Source/PLCrashReportBinaryImageInfo.m [hunk 16]
```diff
@@ -61,8 +61,8 @@ - (id) initWithCodeType: (PLCrashReportProcessorInfo *) processorInfo
 
     _baseAddress = baseAddress;
     _imageSize = size;
-    _imageName = [name retain];
-    _processorInfo = [processorInfo retain];
+    _imageName = name;
+    _processorInfo = processorInfo;
```
Comment (MatkovIvan) [inline Source/PLCrashReportBinaryImageInfo.m:65]: Add explicit modifiers

File: Source/PLCrashReportMachExceptionInfo.m [hunk 17]
```diff
@@ -44,14 +44,9 @@ - (id) initWithType: (uint64_t) type codes: (NSArray *) codes {
         return nil;
     
     _type = type;
-    _codes = [codes retain];
+    _codes = codes;
```
Comment (MatkovIvan) [inline Source/PLCrashReportMachExceptionInfo.m:47]: Add explicit modifiers

File: Source/PLCrashReportMachineInfo.m [hunk 18]
```diff
@@ -61,19 +61,12 @@ - (id) initWithModelName: (NSString *) modelName
     if ((self = [super init]) == nil)
         return nil;
 
-    _modelName = [modelName retain];
-    _processorInfo = [processorInfo retain];
+    _modelName = modelName;
+    _processorInfo = processorInfo;
```
Comment (MatkovIvan) [inline Source/PLCrashReportMachineInfo.m:65]: Add explicit modifiers

File: Source/PLCrashReportProcessInfo.m [hunk 19]
```diff
@@ -61,25 +61,17 @@ - (id) initWithProcessName: (NSString *) processName
...
     
-    _processName = [processName retain];
+    _processName = processName;
     _processID = processID;
-    _processPath = [processPath retain];
-    _processStartTime = [processStartTime retain];
-    _parentProcessName = [parentProcessName retain];
+    _processPath = processPath;
+    _processStartTime = processStartTime;
+    _parentProcessName = parentProcessName;
```
Comment (MatkovIvan) [inline Source/PLCrashReportProcessInfo.m:68]: Add explicit modifiers

File: Source/PLCrashReportRegisterInfo.m [hunk 20]
```diff
@@ -41,17 +41,12 @@ - (id) initWithRegisterName: (NSString *) registerName registerValue: (uint64_t)
     if ((self = [super init]) == nil)
         return nil;
     
-    _registerName = [registerName retain];
+    _registerName = registerName;
```
Comment (MatkovIvan) [inline Source/PLCrashReportRegisterInfo.m:44]: Add explicit modifiers

File: Source/PLCrashReportSignalInfo.m [hunk 21]
```diff
@@ -41,19 +41,13 @@ - (id) initWithSignalName: (NSString *) name code: (NSString *) code address: (u
     if ((self = [super init]) == nil)
         return nil;
     
-    _name = [name retain];
-    _code = [code retain];
+    _name = name;
+    _code = code;
```
Comment (MatkovIvan) [inline Source/PLCrashReportSignalInfo.m:45]: Add explicit modifiers

File: Source/PLCrashReportStackFrameInfo.m [hunk 22]
```diff
@@ -47,15 +47,8 @@ - (id) initWithInstructionPointer: (uint64_t) instructionPointer symbolInfo: (PL
         return nil;
     
     _instructionPointer = instructionPointer;
-    _symbolInfo = [symbolInfo retain];
-    
+    _symbolInfo = symbolInfo;
```
Comment (MatkovIvan) [inline Source/PLCrashReportStackFrameInfo.m:50]: Add explicit modifiers

File: Source/PLCrashReportSymbolInfo.m [hunk 23]
```diff
@@ -52,17 +52,11 @@ - (id) initWithSymbolName: (NSString *) symbolName
     if ((self = [super init]) == nil)
         return nil;
 
-    _symbolName = [symbolName retain];
+    _symbolName = symbolName;
     _startAddress = startAddress;
     _endAddress = endAddress;
```
Comment (MatkovIvan) [inline Source/PLCrashReportSymbolInfo.m:57]: Add explicit modifiers

File: Source/PLCrashReportSystemInfo.m [hunk 24]
```diff
@@ -147,23 +147,15 @@ - (id) initWithOperatingSystem: (PLCrashReportOperatingSystem) operatingSystem
...
     _operatingSystem = operatingSystem;
-    _osVersion = [operatingSystemVersion retain];
-    _osBuild = [operatingSystemBuild retain];
+    _osVersion = operatingSystemVersion;
+    _osBuild = operatingSystemBuild;
     _architecture = architecture;
-    _processorInfo = [processorInfo retain];
-    _timestamp = [timestamp retain];
+    _processorInfo = processorInfo;
+    _timestamp = timestamp;
```
Comment (MatkovIvan) [inline Source/PLCrashReportSystemInfo.m:154]: Add explicit modifiers

File: Source/PLCrashReportTextFormatter.m [hunk 25]
```diff
@@ -182,8 +182,7 @@ + (NSString *) stringValueForCrashReport: (PLCrashReport *) report withTextForma
 
         NSString *incidentIdentifier = @"???";
         if (report.uuidRef != NULL) {
-            incidentIdentifier = (NSString *) CFUUIDCreateString(NULL, report.uuidRef);
-            [incidentIdentifier autorelease];
+            incidentIdentifier = (__bridge_transfer NSString *) CFUUIDCreateString(NULL, report.uuidRef);
```
Comment (MatkovIvan) [inline Source/PLCrashReportTextFormatter.m:185]: NULL -> nil

File: Source/PLCrashReportThreadInfo.m [hunk 26]
```diff
@@ -47,19 +47,13 @@ - (id) initWithThreadNumber: (NSInteger) threadNumber
         return nil;
 
     _threadNumber = threadNumber;
-    _stackFrames = [stackFrames retain];
+    _stackFrames = stackFrames;
     _crashed = crashed;
-    _registers = [registers retain];
+    _registers = registers;
```
Comment (MatkovIvan) [inline Source/PLCrashReportThreadInfo.m:52]: Add explicit modifiers

File: Source/PLCrashReporter.m [hunk 27]
```diff
@@ -54,6 +54,11 @@
     NSLog(@"[PLCrashReporter] " msg, ## args); \
 }
 
+/**
+ * Static field to prevent ARC from freeing resources.
```
Comment (MatkovIvan) [inline Source/PLCrashReporter.m:58]: static here means file visibility without creating symbol. Update comment

File: Source/PLCrashReporter.m [hunk 28]
```diff
@@ -608,24 +613,24 @@ - (BOOL) enableCrashReporterAndReturnError: (NSError **) outError {
...
+            _machServer = [self enableMachExceptionServerWithPreviousPortSet:  &_previousMachPorts
                                                                     callback: &mach_exception_callback
                                                                      context: &signal_handler_context
                                                                        error: outError];
             if (_machServer == nil)
                 return NO;
             
-            /* Acquire references to the autoreleased values */
-            [_machServer retain];
-            [_previousMachPorts retain];
```
Comment (MatkovIvan) [inline Source/PLCrashReporter.m:620]: Add explicit modifiers for this fields

File: Source/PLCrashReporter.m [hunk 29]
```diff
@@ -608,24 +613,24 @@ - (BOOL) enableCrashReporterAndReturnError: (NSError **) outError {
...
-            
             /*
              * MEMORY WARNING: To ensure that our instance survives for the lifetime of the callback registration,
-             * we retain it here. This is necessary to ensure that the Mach exception server instance and previous port set
+             * we keep a reference on self. This is necessary to ensure that the Mach exception server instance and previous port set
              * survive for the lifetime of the callback. Since there's currently no support for *deregistering* a crash reporter,
              * this simply results in the reporter living forever.
              */
-            [self retain];
+            if (pl_crash_reporter != NULL) {
```
Comment (MatkovIvan) [inline Source/PLCrashReporter.m:629]: 1. `nil`
2. Cane we move the check above actual initializing?

File: Source/PLCrashReporter.m [hunk 30]
```diff
@@ -858,17 +863,17 @@ - (id) initWithApplicationIdentifier: (NSString *) applicationIdentifier appVers
...
 
     /* Save the configuration */
-    _config = [configuration retain];
-    _applicationIdentifier = [applicationIdentifier retain];
-    _applicationVersion = [applicationVersion retain];
-    _applicationMarketingVersion = [applicationMarketingVersion retain];
+    _config = configuration;
+    _applicationIdentifier = applicationIdentifier;
+    _applicationVersion = applicationVersion;
+    _applicationMarketingVersion = applicationMarketingVersion;
```
Comment (MatkovIvan) [inline Source/PLCrashReporter.m:869]: Add explicit modifiers

File: Tests/PLCrashAsyncMachOImageTests.m [hunk 31]
```diff
@@ -48,8 +48,9 @@ @implementation PLCrashAsyncMachOImageTests
 - (void) setUp {
     /* Fetch our containing image's dyld info */
     Dl_info info;
-    STAssertTrue(dladdr([self class], &info) > 0, @"Could not fetch dyld info for %p", [self class]);
-
+    void *classRef = (__bridge_retained void *)([self class]);
```
Comment (MatkovIvan) [inline Tests/PLCrashAsyncMachOImageTests.m:51]: No need to transfer ownership here. Why not just `__bridge`? See [an article about it](https://medium.com/@notestomyself/bridge-vs-bridge-transfer-vs-bridge-retained-f84e6b6406d1)

File: Source/PLCrashMachExceptionPortSet.m (hunk 14 above)
Comment (annakocheshkova) [reply Source/PLCrashMachExceptionPortSet.m:116]: We don't know either. This signature was updated for compliance with `NSFastEnumeration` protocol: 
image

Comment (annakocheshkova) [reply Source/PLCrashMachExceptionPortSet.m:116]: I guess this means we can leave only one option here...

File: Source/PLCrashReporter.m [hunk 32]
```diff
@@ -608,24 +613,22 @@ - (BOOL) enableCrashReporterAndReturnError: (NSError **) outError {
...
              * MEMORY WARNING: To ensure that our instance survives for the lifetime of the callback registration,
-             * we retain it here. This is necessary to ensure that the Mach exception server instance and previous port set
+             * we keep a reference on self. This is necessary to ensure that the Mach exception server instance and previous port set
              * survive for the lifetime of the callback. Since there's currently no support for *deregistering* a crash reporter,
              * this simply results in the reporter living forever.
+             * We do not check pl_crash_reporter nullability here, because we already ensured
+             * that this is a single instance.
              */
-            [self retain];
+            pl_crash_reporter = self;
```
Comment (landonf) [inline Source/PLCrashReporter.m:631]: `CFBridgingRetain()` (with a corresponding `CFBridgingRelease()`) would be more appropriate here, and would avoid introducing additional global singleton state.

File: Tests/PLCrashAsyncMachOImageTests.m [hunk 33]
```diff
@@ -48,8 +48,9 @@ @implementation PLCrashAsyncMachOImageTests
 - (void) setUp {
     /* Fetch our containing image's dyld info */
     Dl_info info;
-    STAssertTrue(dladdr([self class], &info) > 0, @"Could not fetch dyld info for %p", [self class]);
-
+    void *classRef = (__bridge void *)([self class]);
+    STAssertTrue(dladdr(classRef, &info) > 0, @"Could not fetch dyld info for %p", [self class]);
+    CFBridgingRelease(classRef);
```
Comment (landonf) [inline Tests/PLCrashAsyncMachOImageTests.m:53]: __bridge doesn't incur a retain, and the `-class` reference is a borrowed one; it should not be explicitly released.

File: Source/PLCrashReporter.m (hunk 32 above)
Comment (MatkovIvan) [reply Source/PLCrashReporter.m:631]: Good to see you here. Thank you for following the changes

Comment (annakocheshkova) [reply Source/PLCrashReporter.m]: @landonf I don't think `CFBridgingRelease()` is necessary here. Prior to our changes there was a single `[self retain]` here - without any explicit releasing (and the explanation was in the comment). 
So I think it makes sense to just replace it with `CFBridgingRetain()`.

**Pull request merged**